### PR TITLE
FilterHelper.py: Rewrite filters

### PR DESCRIPTION
**Pull request by Utkarsh1308**
Change the filters input from list to dict

Fixes https://github.com/coala/coala/issues/5398

**For short term contributors:** we understand that getting your commits well
defined like we require is a hard task and takes some learning. If you
look to help without wanting to contribute long term there's no need
for you to learn this. Just drop us a message and we'll take care of brushing
up your stuff for merge!

### Checklist

- [x] I read the [commit guidelines](http://coala.io/commit) and I've followed
      them.
- [x] I ran coala over my code locally. (*All commits have to pass
      individually.* It is not sufficient to have "fixup commits" on your PR,
      our bot will still report the issues for the previous commit.) You will
      likely receive a lot of bot comments and build failures if coala does not
      pass on every single commit!

After you submit your pull request, **DO NOT click the 'Update Branch' button.**
When asked for a rebase, consult [coala.io/rebase](https://coala.io/rebase)
instead.

Please consider helping us by reviewing other peoples pull requests as well:

- pick up any PR at <https://coala.io/review>
- review it (check <https://coala.io/reviewing> for more info)
- if you are sure that it needs work, use `corobo mark wip <URL>` to get it out
  of the review queue.

The more you review, the more your score will grow at coala.io and we will
review your PRs faster!

### CODE REVIEW

**Comment by ksdme** [comment]
https://github.com/coala/coala/blob/master/coalib/coala.py#L69 might require some patch work too.

**Comment by ksdme** [comment]
Would you consider doing this in stages? First let us transition this into `tuple` based structures and then worry about introducing dicts (OrderedDicts) or class based filters.

File: coalib/parsing/FilterHelper.py [hunk 1]
```diff
@@ -75,8 +75,11 @@ def apply_filters(filters, bears=None, sections=None):
         items = sections
         applier = _apply_section_filter
 
-    for filter in filters:
-        filter_name, *filter_args = filter
+    for filter_name, *filter_args in filters.items():
+        for key in filter_args:
+            if isinstance(key, list):
+                filter_args = [
+                    item for sublist in filter_args for item in sublist]
```
[the remark this answers is not in the document]
Comment (Utkarsh1308) [reply coalib/parsing/FilterHelper.py]: To convert a list of lists into a list. --> [['C', 'java']] ==> ['C', 'java']

File: coalib/parsing/FilterHelper.py [hunk 2]
```diff
@@ -75,8 +75,11 @@ def apply_filters(filters, bears=None, sections=None):
         items = sections
         applier = _apply_section_filter
 
-    for filter in filters:
-        filter_name, *filter_args = filter
+    for filter_name, *filter_args in filters.items():
```
[the remark this answers is not in the document]
Comment (Utkarsh1308) [reply coalib/parsing/FilterHelper.py]: In the original code filter_args gave a list instead of a single variable. (['syntax'] instead of syntax)
On using filter_args I got a single variable and *filter_args gave a list so I sticked with the original result because I didn't know if this change could give an error.

File: coalib/parsing/FilterHelper.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Utkarsh1308) [reply coalib/parsing/FilterHelper.py]: This would give the following result 
```
language
('c', 'java')
can_fix
('s', 'y', 'n', 't', 'a', 'x')
section_tags
('s', 'a', 'v', 'e')
```
instead of what I am aiming for 
```
language
['c', 'java']
can_fix
['syntax']
section_tags
['save']
```
I wrote this extra code to change the filter args from 
```
[['c', 'java']]
['syntax']
['save']
```
to 
```
['c', 'java']
['syntax']
['save']
```

File: coalib/parsing/FilterHelper.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ksdme) [reply coalib/parsing/FilterHelper.py]: `filters` is a list right?

[the remark this answers is not in the document]
Comment (Utkarsh1308) [reply coalib/parsing/FilterHelper.py]: The original code takes filter as a list but according to my understanding the issue aims to change filters from a list to a dict as an argument. I might have misunderstood the issue.

File: coalib/parsing/FilterHelper.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ksdme) [reply coalib/parsing/FilterHelper.py]:
```py
>>> filters = [['langauge', 'c', 'java'], ['can_fix', 'syntax']]
>>> dict_filters = {}
>>> 
>>> for filter_name, *filter_args in filters:
...   filter_args = tuple(filter_args)
...   dict_filters[filter_name] = filter_args
... 
>>> 
>>> dict_filters
{'langauge': ('c', 'java'), 'can_fix': ('syntax',)}
>>> 
```

What am I doing wrong here?

[the remark this answers is not in the document]
Comment (Utkarsh1308) [reply coalib/parsing/FilterHelper.py]: I changed filters from a list into a dict.
The new filters is 
```
{'language': ['c', 'java'], 'can_fix': 'syntax', 'section_tags': 'save'}
```

[the remark this answers is not in the document]
Comment (ksdme) [reply coalib/parsing/FilterHelper.py]: Ok, so my understanding is you have already performed partial conversion to a dict based filter structure. A few remarks here:
- Use OrderedDicts, filtering order depends is important here.
- Use tuples for holding args, not lists.
- Do not partially convert the filters, you should rather have all the filter args as tuples. `{'language': ('c', 'java'), 'can_fix': ('syntax',), 'section_tags': ('save',)}` otherwise it will force all the other unrelated methods to handle this case.

[the remark this answers is not in the document]
Comment (ksdme) [reply coalib/parsing/FilterHelper.py]: I would suggest doing this step wise to understand it better. Make everything work with tuple right now and then introduce dicts.

File: coalib/parsing/FilterHelper.py [hunk 3]
```diff
@@ -60,8 +60,8 @@ def apply_filters(filters, bears=None, sections=None):
 
     :param filters:
         List of args based on ``bears`` has to be filtered. For example:
```
[the remark this answers is not in the document]
Comment (Utkarsh1308) [reply coalib/parsing/FilterHelper.py]: What should be changed?

File: coalib/parsing/FilterHelper.py [hunk 4]
```diff
@@ -123,10 +126,20 @@ def collect_filters(args, arg_list=None, arg_parser=None):
...
         arg_parser = default_arg_parser() if arg_parser is None else arg_parser
         args = arg_parser.parse_args(arg_list)
 
-    return getattr(args, 'filter_by', None) or []
+    filters = getattr(args, 'filter_by', None) or []
+    filters = list_to_dict(filters)
+    return filters
+
+
+def list_to_dict(filters):
```
Comment (ksdme) [inline coalib/parsing/FilterHelper.py]: Please name this appropriately, list_to_dict does not explain its purpose.

File: coalib/parsing/FilterHelper.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (li-boxuan) [reply coalib/parsing/FilterHelper.py]: It is not a list of args anymore, isn't it?

File: coalib/parsing/FilterHelper.py [hunk 5]
```diff
@@ -123,10 +125,20 @@ def collect_filters(args, arg_list=None, arg_parser=None):
...
+    items = {}
+    for filter in filters:
+        key, value = filter[0], filter[1:]
```
Comment (ksdme) [inline coalib/parsing/FilterHelper.py]: `value` should be tuple. I have already told you this. Also please rename appropriately to say `filter_name`, `args`.

File: coalib/parsing/FilterHelper.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Utkarsh1308) [reply coalib/parsing/FilterHelper.py]: Oh sorry foolish me.

File: coalib/parsing/FilterHelper.py (hunk 4 above)
Comment (Utkarsh1308) [reply coalib/parsing/FilterHelper.py]: I asked a maintainer regarding what name should I give to this function and he suggested this. Please suggest a better name.

File: coalib/parsing/FilterHelper.py [hunk 6]
```diff
@@ -75,8 +75,11 @@ def apply_filters(filters, bears=None, sections=None):
         items = sections
         applier = _apply_section_filter
 
-    for filter in filters:
-        filter_name, *filter_args = filter
+    for filter_name, *filter_args in filters.items():
+        for filter_arg in filter_args:
+            if isinstance(filter_arg, list):
+                filter_args = [
+                    item for sublist in filter_args for item in sublist]
```
[the remark this answers is not in the document]
Comment (Utkarsh1308) [reply coalib/parsing/FilterHelper.py]: The code does not comply to pep8 hence the new line.

File: coalib/parsing/FilterHelper.py (hunk 4 above)
Comment (ksdme) [reply coalib/parsing/FilterHelper.py]: `flat_filter_to_dict` or `filter_vector_to_dict` maybe? This is not a very generic function hence that is not a very appropriate name.

Comment (ksdme) [reply coalib/parsing/FilterHelper.py]: And have you discussed anything about the tuple/list situation concerning to args?

Comment (Utkarsh1308) [reply coalib/parsing/FilterHelper.py]: I haven't yet. I'll try to make it work with tuples.

File: coalib/parsing/FilterHelper.py [hunk 7]
```diff
@@ -59,9 +60,9 @@ def apply_filters(filters, bears=None, sections=None):
     and runs filter in bear filtering mode.
 
     :param filters:
-        List of args based on ``bears`` has to be filtered. For example:
-        ``[['language', 'c', 'java'], ['can_fix', 'syntax'],
-        ['section_tags', 'save']]``
+        Dict of args based on ``bears`` has to be filtered. For example:
+        ``{'language': ('c', 'java'), 'can_fix': 'syntax',
```
Comment (ksdme) [inline coalib/parsing/FilterHelper.py]: Docs need update, please change to `('syntax',)`; https://github.com/coala/coala/pull/5707/commits/721aabda8accf955ba01be090339b405f0dfbc10#diff-0ba65c8ebc811938f8e34a547109a908R105

File: coalib/results/result_actions/GeneratePatchesAction.py [hunk 8]
```diff
@@ -41,7 +41,7 @@ def filter_bears(language):
     :param language: The language to filter with.
     :return:         A list of bears.
     """
-    return list(apply_filters([['language', language]], None)[0]['cli'])
+    return list(apply_filters({'language': language}, None)[0]['cli'])
```
Comment (ksdme) [inline coalib/results/result_actions/GeneratePatchesAction.py]: Needs to be `(language,)` right?, seems like we need better tests.

File: tests/parsing/FilterHelperTest.py [hunk 9]
```diff
@@ -51,8 +51,8 @@ def test_apply_filter(self):
         self.assertIsNotNone(apply_filter_result)
 
     def test_apply_filters(self):
-        apply_filters_result = apply_filters([('language', 'C', 'Python'),
-                                              ('can_fix', 'syntax')])
+        apply_filters_result = apply_filters({'language': ('C', 'Python'),
+                                              'can_fix': 'syntax'})
```
Comment (ksdme) [inline tests/parsing/FilterHelperTest.py]: How is this test passing when we need to have args for `can_fix` as a tuple rather than as a string?

File: coalib/parsing/FilterHelper.py [hunk 10]
```diff
@@ -75,8 +76,11 @@ def apply_filters(filters, bears=None, sections=None):
         items = sections
         applier = _apply_section_filter
 
-    for filter in filters:
-        filter_name, *filter_args = filter
+    for filter_name, *filter_args in filters.items():
+        for filter_arg in filter_args:
+            if isinstance(filter_arg, tuple):
```
Comment (ksdme) [inline coalib/parsing/FilterHelper.py]: How is this code still relevant? Can't `filter_vector_to_dict` take care of all this?

Comment (Utkarsh1308) [reply coalib/parsing/FilterHelper.py]: I thought filter_vector_to_dict was supposed to change the input to a dict. And this code is supposed to parse through that input.

File: coalib/parsing/FilterHelper.py [hunk 11]
```diff
@@ -25,7 +26,7 @@ def apply_filter(filter_name, filter_args, all_bears=None):
     :param filter_args:
         Arguments of the filter to be passed in.
         For example:
-        ``['c', 'java']``
+        ``('c', 'java',)``
```
Comment (ksdme) [inline coalib/parsing/FilterHelper.py]: To be consistent you should drop the trailing comma.

File: coalib/parsing/FilterHelper.py [hunk 12]
```diff
@@ -59,9 +60,9 @@ def apply_filters(filters, bears=None, sections=None):
     and runs filter in bear filtering mode.
 
     :param filters:
-        List of args based on ``bears`` has to be filtered. For example:
-        ``[['language', 'c', 'java'], ['can_fix', 'syntax'],
-        ['section_tags', 'save']]``
+        Dict of args based on ``bears`` has to be filtered. For example:
+        ``{'language': ('c', 'java'), 'can_fix': ('syntax',),
```
Comment (ksdme) [inline coalib/parsing/FilterHelper.py]: Since we are using `OrderedDict`'s you should update the docs to show that. Please go over all the docs and update this.

File: coalib/parsing/FilterHelper.py [hunk 13]
```diff
@@ -1,10 +1,11 @@
 from coalib.parsing.InvalidFilterException import InvalidFilterException
 from coalib.parsing.filters import available_filters
 from coalib.parsing.DefaultArgParser import default_arg_parser
+from collections import OrderedDict
```
Comment (ksdme) [inline coalib/parsing/FilterHelper.py]: Please move this import to the top and give a space to confirm to our standards. Standard library imports should be topmost. Refer https://github.com/coala/coala/blob/d885c4ad2729282cf43083e08be041cf70e458bf/coalib/io/FileFactory.py

File: coalib/parsing/FilterHelper.py [hunk 14]
```diff
@@ -123,10 +123,20 @@ def collect_filters(args, arg_list=None, arg_parser=None):
...
         arg_parser = default_arg_parser() if arg_parser is None else arg_parser
         args = arg_parser.parse_args(arg_list)
 
-    return getattr(args, 'filter_by', None) or []
+    filters = getattr(args, 'filter_by', None) or []
+    filters = filter_vector_to_dict(filters)
+    return filters
+
+
+def filter_vector_to_dict(filters):
```
Comment (ksdme) [inline coalib/parsing/FilterHelper.py]: Please add docs to this function.

**Comment by ksdme** [comment]
The test failing is `tests/coalaTest.py::coalaTest::test_coala_with_color`, you can ignore that, it happens sometimes.

File: coalib/parsing/FilterHelper.py [hunk 15]
```diff
@@ -59,9 +61,9 @@ def apply_filters(filters, bears=None, sections=None):
     and runs filter in bear filtering mode.
 
     :param filters:
-        List of args based on ``bears`` has to be filtered. For example:
-        ``[['language', 'c', 'java'], ['can_fix', 'syntax'],
-        ['section_tags', 'save']]``
+        OrderedDict of args based on ``bears`` has to be filtered. For example:
```
Comment (ksdme) [inline coalib/parsing/FilterHelper.py]: `OrderedDict` of filters, right?

**Comment by jayvdb** [comment]
@gitmate-bot rebase

**Comment by jayvdb** [comment]
@gitmate-bot rebase

**Comment by jayvdb** [comment]
@gitmate-bot rebase

**Comment by jayvdb** [comment]
ack 7211a18

**Comment by jayvdb** [comment]
@gitmate-bot ff

**Comment by jayvdb** [comment]
@gitmate-bot ff
